src/main/java/com/algorand/algosdk/transaction/AtomicTransactionComposer.java

```diff
@@ -0,0 +1,439 @@
+     */
+    public List<String> submit(AlgodClient client) throws Exception {
+        if (this.status.compareTo(Status.SUBMITTED) > 0)
+            throw new IllegalArgumentException("Atomic Transaction Composer cannot submit committed transaction");
+
+        this.gatherSignatures();
+
+        List<byte[]> encodings = new ArrayList<>();
+        int lengthEncoded = 0;
+        for (SignedTransaction stx : this.signedTxns) {
+            byte[] temp = Encoder.encodeToMsgPack(stx);
+            encodings.add(temp);
+            lengthEncoded += temp.length;
+        }
+        ByteBuffer bf = ByteBuffer.allocate(lengthEncoded);
+        for (byte[] subEncode : encodings)
+            bf.put(subEncode);
+        byte[] encoded = bf.array();
+        Response<PostTransactionsResponse> rPost = client.RawTransaction().rawtxn(encoded).execute();
+
+        if (!rPost.isSuccessful())
+            throw new Exception("transaction should be submitted successfully");
+
+        this.status = Status.SUBMITTED;
+        return this.getTxIDs();
+    }
+
+    /**
+     * Send the transaction group to the network and wait until it's committed to a block. An error
+     * will be thrown if submission or execution fails.
+     * <p>
+     * The composer's status must be SUBMITTED or lower before calling this method, since execution is
+     * only allowed once. If submission is successful, this composer's status will update to SUBMITTED.
+     * If the execution is also successful, this composer's status will update to COMMITTED.
+     * <p>
+     * Note: a group can only be submitted again if it fails.
+     *
+     * @return If the execution is successful, resolves to an object containing the confirmed round for
+     * this transaction, the txIDs of the submitted transactions, and an array of results containing
+     * one element for each method call transaction in this group. If a method has no return value
+     * (void), then the method results array will contain null for that method's return value.
+     */
+    public ExecuteResult execute(AlgodClient client, int waitRounds) throws Exception {
+        if (this.status == Status.COMMITTED)
+            throw new IllegalArgumentException("status shows this is already committed");
+        if (waitRounds < 0)
+            throw new IllegalArgumentException("wait round for execute should be non-negative");
+        this.submit(client);
+
+        int firstMethodCallIndex = -1;
+        for (int i = 0; i < this.signedTxns.size(); i++) {
+            if (this.signedTxns.get(i).tx.type.equals(Transaction.Type.ApplicationCall)) {
+                firstMethodCallIndex = i;
+                break;
+            }
+        }
+        firstMethodCallIndex = (firstMethodCallIndex == -1) ? 0 : firstMethodCallIndex;
+
+        PendingTransactionResponse txInfo =
+                Utils.waitForConfirmation(client, this.signedTxns.get(firstMethodCallIndex).transactionID, waitRounds);
```

Comment:
It's only useful to wait on `firstMethodCallIndex` if you reuse the output of `Utils.waitForConfirmation` instead of calling `client.PendingTransactionInformation` for the first method call.

Also, right now `firstMethodCallIndex` is actually the index of the first app call transaction, not necessarily the first method call, since non-ABI app calls can be added with `addTransaction`.

Reply:
It sounds like it is an unnecessary move to use `firstMethodCallIndex`, assuming that any transaction in this transaction group is committed to the network at the same round.

Given the knowledge that we are holding all transaction id's, I suppose this is simplified to using first transaction id in signed transaction list?

Reply:
Since the group is atomic, we can pass any of the TxIDs to `Utils.waitForConfirmation` to wait for the whole group to be committed. This function returns a `PendingTransactionResponse`, and since the execute function needs to obtain a `PendingTransactionResponse` for every method call in the group to examine logged values, the implementation could potentially reduce one API call if it passes a method call TxID to `Utils.waitForConfirmation`.

But this is just an optimization, so it's not necessary to implement. The current implementation is definitely acceptable.